Initial codification of robot upgrades

Veracity0 opened the pull request:
This is the PR for major refactoring/cleanup/tests of YouRobotManager.

1) All robot parts and CPU upgrades are now in enum RobotUpgrade.
2) When you visit the Reassembly Center and purchase parts or CPU upgrades, we now log a more meaningful message:
Inspecting Top Attachment options at the Reassembly Station.
Installing Solar Panel as your Top Attachment for 5 scrap.
Inspecting Right Arm options at the Reassembly Station.
Installing Omni-Claw as your Right Arm for 15 scrap.
Inspecting CPU Upgrade options at the Reassembly Station.
Upgrading your CPU with Topology Grid for 50 energy.
3) Similarly for visiting the Statbot 5000
Spending 21 energy to upgrade Moxie by 5 points.
You gain 1035 Sarcasm
You gain some Moxie points!
You gain a Level!
4) When you install the Topology Grid, it actually gives you Torso Awareness. Therefore, we do not need to check for that upgrade in KoLCharacter.isTorsoAware
5) Tests:
Can parse avatar from char sheet and set appropriate properties
Can parse avatar from char pane and set appropriate properties
Can parse avatar from visit to Reassembly Station and set appropriate properties
Can log pretty messages when perform visits and actions at the Reassembly Station
Can log pretty message when activate Statbot 5000
Can detect the current energy cost when visit Statbot 5000
Increments usage count when successfully use Statbot 5000
Adds/loses combat skills as appropriate when swap parts. (*** disabled; existing bug)
When you swap out a part that lets you equip items, any equipped item is dropped
When you visit the Reassembly Station, we can parse CPU upgrades
When you buy a CPU upgrade, set of known upgrades is updated
Parses Statbot usages even if activation failed.
When you swap out a Bird Cage, current familiar is dropped.
6) Bug fix: We parse Statbot 5000 when we visit and derive the number of activations from the energy cost. Yet, after we activate it, we do not increment the activation count. That can't be right; visiting Statbot 5000 again immediately will update. I made the activation simply parse the responseText and derive it again, just as on the initial visit.
7) Bug fix: taltimir commented that Quest.TRAPPER did not advance after Groarbot. There were four other quests which similarly checked for non-standard bosses, but not You, Robot. I added You, Robot bosses to all of those checks. This is untested, since I never think to look at quest progress and I am long past all of these - except for the Sorceress. Man. That's a huge if-then-elseif chain of monster name string comparisons in QuestManager.updateQuestData(). Seems like it could be a switch statement, just as it is in QuestManager.updateQuestFightLost().
8) Given all those tests, we no longer use any of the legacy properties, although they are all set compatibly.

Veracity0 commented:
To carry on a comment from my previous PR:

I welcome comments even before I have declared this "ready to review". I will pay attention and - unless I have already freed the king and can no longer test things things that require me to be in-path - address them.

Something I have already noticed:

On You, Robot, you cannot eat - EXCEPT for magic sausages. That's a little annoying; if I grey out the Consume button for eating, because you "can't eat", I can't creat/eat magical sausages except via the gCLI or Relay Browser. Perhaps I should NOT grey it out, but not let you queue (or attempt to  consume) anything except that food?

And it was pointed out that my avater image handling could just as well take a List, rather than a (possibly 0-length) array? Sure. I was adapting existing code. I could have adapted it better.

Here are my current thought on how to proceed with this:

- Implement YouRobotManager.registerRequest

- Write tests!
--- YouRobotRequest.parseAvatar
---- Sets preferences and avatar
---- Listener fires if the Avatar changes
---- Listener doesn't fire if the Avatar doesn't change
--- YouRobotRequest.parseCPUUpgrades
--- more

- Removing EQUIP gear:
-- remove current {hat, weapon, offhand, pants, familiar}
-- fire (hats), (weapons), (offhand), (pants), (familiar) to alert GearChanger

- Adding EQUIP gear:
-- fire (hats), (weapons), (offhand), (pants), (familiar) to alert GearChanger

- Removing COMBAT gear
-- remove Available Combat Skill

- Adding COMBAT gear
-- add Available Combat Skill

- Add or Remove PASSIVE gear
-- KoLCharacter.recalculate adjustments

- Adding CPU upgrades:
-- fire (shirts) or (potions), as appropriate
-- KoLCharacter.recalculate adjustments

Improvements:

- parseAvatar sets "current" variable
-- parseAvatar only updates preferences if they have changed
-- parseAvatar only updates avatar if it have changed
-- Fire listeners as appropriate: (hats), (pants), (weapons), (offhands), (shirts)

- parseCPUUpgrades sets "current" variable
-- parseCPUUpgrades only updates preference if upgrades have changed
-- Fire listeners as appropriate: (shirts), (potions)

- GearChanger "hats" dropdowns listens for (hats)
-- GearChanger "weapons" dropdowns listens for (weapons)
-- GearChanger "offhands" dropdowns listens for (offhands)
-- GearChanger "pants" dropdowns listens for (pants)
-- GearChanger "shirts" dropdowns listens for (shirts)
-- GearChanger "familiars" dropdowns listens for (familiars)

- Item Manager "potions" panel listens for (potions)

Veracity0 commented:
This is ready for review.

I'm going to run my turns, culminating in freeing the king.
On the way, I will collect HTML for swapping out a Bird Cage (containing a familiar).
I will write a test confirming that doing so leaves you with no current familiar.

I have more bugs to fix and code cleanup to make, but as of this PR, we no longer depend on the legacy properties for our internal operation - although we set them compatibly, based on our internal configuration. All the tests check for correctness based on the laegacy properties, rather than by looking at the internal data structures. Black-ish Box testing. :)

I will do aftercore today and then tomorrow, do more aftercore and ascend into my next run. That gives you about 24 hours to look at this and comment and for me to address your comments.

I will follow up here with a comment describing everything else I am still thinking of looking into.

midgleyc commented on test/net/sourceforge/kolmafia/request/ScrapheapRequestTest.java:
```diff
@@ -65,7 +66,10 @@ public void parseCPUUpgrades() throws IOException {
+    Arrays.sort(expected);
+
     var actual = Preferences.getString("youRobotCPUUpgrades").split(",");
+    Arrays.sort(actual);
```
Instead of sorting the arrays, you could use hamcrest matchers:
```
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.arrayContainingInAnyOrder;

assertThat(actual, arrayContainingInAnyOrder(expected));
```

Given that the number of expected matches is so large I'm not sure that reads any better than sorting them, but it could be nicer in other cases with fewer expected matches.

Veracity0 replied:
OK.

Veracity0 replied:
Done.

midgleyc commented on test/net/sourceforge/kolmafia/session/YouRobotManagerTest.java:
```diff
@@ -0,0 +1,354 @@
+    YouRobotManager.reset();
+    KoLCharacter.setAvatar("");
+    ChoiceManager.lastChoice = 0;
+    ChoiceManager.lastChoice = 0;
```
duplication, unnecessary?

Veracity0 replied:
Oops. Supposed to be lastChoice and lastDecision. Will fix.

Veracity0 replied:
Done.

midgleyc commented on src/net/sourceforge/kolmafia/session/YouRobotManager.java:
```diff
@@ -16,92 +22,528 @@
+        currentParts.put(part, upgrade);
+        // Set the legacy properties for use by scripts
+        Preferences.setInteger("youRobot" + part.getSection(), upgrade.getIndex());
+        // *** Fire listeners?
```
presumably these are both TODO? Combat skills has a disabled test.

Veracity0 replied:
Right. I am about to post my list of TODOs. Firing listeners is pointless unless GUI elements are listening for them. :)

Veracity0 commented:
Here is my TODO list for next PR to be looked at starting with next run:

Bugs:

- On You, Robot, you cannot eat - EXCEPT for magic sausages.
Since I disable the Consume button if you cannot eat, I cannot queue/create/eat them.
I need to rethink that, since having a fullness limit of 0 does not mean you cannot eat 0 fullness foods.

- midgleyc pointed out that my avater image handling could just as well take a List, rather than a (possibly 0-length) array?

- taltimir: there is a maximizer issue where it thinks muscle provides hp in you, robot. it does not.

Improvements:

- Removing EQUIP gear:
remove current {hat, weapon, offhand, pants, shirts, familiar}
fire (hats), (weapons), (offhand), (pants), (shirts), (familiar) as appropriate

- Changing EQUIP gear:
fire (hats), (weapons), (offhand), (pants), (familiar) as appropriate

- GearChanger "hats" dropdowns listens for (hats)
GearChanger "weapons" dropdowns listens for (weapons)
GearChanger "offhands" dropdowns listens for (offhands)
GearChanger "pants" dropdowns listens for (pants)
GearChanger "shirts" dropdowns listens for (shirts)
GearChanger "familiars" dropdowns listens for (familiars)

- Removing COMBAT gear
remove Available Combat Skill
(uncomment disabled test)

- Adding COMBAT gear
add Available Combat Skill
(uncomment disabled test)

- Add or Remove PASSIVE gear
KoLCharacter.recalculate adjustments

- Adding CPU upgrades:
fire (shirts) or (potions), as appropriate
KoLCharacter.recalculate adjustments

- ItemManager potions panel listens for (potions)

Tests:

- Verify that the various listeners get poked as appropriate.
Need a dummy object which can provide an update method and register for  NamedListeners

Veracity0 commented:
I have addressed the code review comments. I am about to ascend, so I am going to approve this.

My first act upon ascension will be to collect a DEBUG log for visiting the Reassembly Station with only 3 parts and verify that we parse the avatar correctly. I'll write a test and submit the HTML and test into a new PR.

My initial PR comment will list the bugs/features/improvements I will be investigating.
